CRM/Report/Form.php

```diff
@@ -5084,6 +5076,12 @@ protected function setOutputMode() {
     if (isset($this->_params['task'])) {
       unset($this->_params['task']);
     }
+
+    if (!empty($this->_outputMode) && empty($this->_outputHandler) && !in_array($this->_outputMode, ['save', 'copy', 'group'])) {
+      $outputHandlerFactory = new \Civi\Report\OutputHandlerFactory();
+      $outputHandlerClass = $outputHandlerFactory->getOutputHandlerClass($this->_outputMode);
+      $this->_outputHandler = new $outputHandlerClass($this);
```

Comment:
I'm not sure about this line. Typically a factory returns an object, for a couple reasons. One is to allow for singletons, or caching. Another is simply that the consumer of the factory shouldn't be the one creating objects. Another is the factory might choose to delegate, e.g. to the container (in a way this is being done with hooks, but is still limiting what it can return - I'm not sure if there's already an example in civi somewhere of e.g. symfony-style dependency injection).

Also this isn't giving very much information to the factory in order to decide, just the outputMode, and the hook isn't even being told that. I'd probably pass $this as a parameter to the factory constructor.

This may go beyond the immediate need, but it seems like it might be hard to change later. Especially the hook.

My two cents.

Reply:
@demeritcowboy I have done this now seems to make more sense

Reply:
Thanks will take a look.